### ISSUE TITLE
add help text to link-in-text-block rule

### ISSUE THREAD

**Issue by shanisebarona** [dequelabs/axe-core#3440]
Add help text with more context to link-in-text-block rule. 

Suggestion: Sufficient color contrast ratio of 3:1 is necessary to distinguish link text color from surrounding text color.

**Comment by WilcoFiers** [comment]
Thank you for the suggestion. I think this rule need some work TBH. The W3C took the position a while back that a 3:1 contrast between the link and it surrounding text would be sufficient. That's not how we understood this criterion when we wrote the rule. Now that a 3:1 contrast ratio is explicitly permitted I think the rule needs some revision.

Out of curiosity, this rule is experimental and so isn't on by default. What are you running axe-core with that has this rule enabled?

**Comment by shanisebarona** [comment]
The rule is currently in Accessibility for Insights for Web as a "needs review" check.

**Comment by DaveTryon** [comment]
@WilcoFiers, as discussed offline, we're picking this up with the following proposed approach:
1. Replace the current link-in-text check with 4 pseudoClass checks--in each case, fall back to the non-pseudoClass colors if colors can't be found in the pseudoClass
  - `link-in-text-block-link` mirrors `link-in-text-block` but checks the `link` pseudoClass
  - `link-in-text-block-hover` mirrors `link-in-text-block` but checks the `hover` pseudoClass
  - `link-in-text-block-visited` mirrors `link-in-text-block` but checks the `visited` pseudoClass
  - `link-in-text-block-active` mirrors `link-in-text-block` but checks the `active` pseudoClass
2. Add the new checks to 'all' array of the `link-in-text-block` rule
3. Update tests appropriately

The net result should be that the `link-in-text-block` rule is able to account for each of the pseudoClass types, then provide a meaningful answer.

Two observations on this approach:
1. Looking at https://www.w3.org/WAI/WCAG21/Understanding/use-of-color, a 3:1 contrast ratio ([G183](https://www.w3.org/WAI/WCAG21/Techniques/general/G183)) is just 1 of 4 possible ways to satisfy the requirement. [G14](https://www.w3.org/WAI/WCAG21/Techniques/general/G14) would satisfy the requirement equally well. If someone were to adopt [G14](https://www.w3.org/WAI/WCAG21/Techniques/general/G14, this rule would essentially be a false positive.
2. A very strict reading of https://www.w3.org/WAI/WCAG21/Understanding/use-of-color might require that we have a 3:1 contrast ratio between **each pairing** of pseudoClass options. We're not currently planning on checking for this, but only checking that the colors for each pseudoClass provide a 3:1 contrast ratio against the colors of the surrounding text block. In an extreme case, imagine that someone specified the same colors for each pseudoClass. As long as the colors had sufficient contrast to the surrounding text, the rule would pass (even though the link colors would never change in response to user action).

Please feel free to comment on the proposed approach and/or the mentioned observations.

**Comment by DaveTryon** [comment]
@WilcoFiers, I've some experimentation with pseudoclasses, and have determined that there are 3 'tiers' where colors can be set, and we'll need to account for this in the revised rule. The tiers are:
Tier | pseudoClasses | Description
---  | --- | ---
High | `hover` or `active` (mutually exclusive) | Classes that reflect the user's _current_ interaction
Middle | `link` or `visited` (mutually exclusive) | Classes that reflect the user's _previous_ interactions
Low | base class or containing text block | Classes that are agnostic of user interaction

When we expand this matrix, we end up with the following checks:
Check | Order of checking colors
--- | ---
`link-active-visited-text-block` | `active`, `visited`, base class, text block
`link-active-link-text-block` | `active`, `link`, base class, text block
`link-hover-visited-text-block` | `hover`, `visited`, base class, text block
`link-hover-link-text-block` | `hover`, `link`, base class, text block
`link-visited-text-block` | `visited`, base class, text block
`link-link-text-block` | `link`, base class, text block

Of course, this still only covers [G183](https://www.w3.org/WAI/WCAG21/Techniques/general/G183) and ignores [G14](https://www.w3.org/WAI/WCAG21/Techniques/general/G14)...

**Comment by DaveTryon** [comment]
I just figured out that G14 is already handled in [elementIsDistinct](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/color/element-is-distinct.js#L27), so please ignore my comments on that. There's also a `focus` pseudoClass that curiously respects `link` but not `visited`

**Comment by DaveTryon** [comment]
We've hit a snag with this. We revised the code to use [window.getComputedStyle](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle), but we've discovered that the API intentionally disregards pseudoclasses. In fact, we even found a [chromium test](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/css/css-pseudo/highlight-pseudos-currentcolor-visited-computed-001.html;l=37) that will fail if this behavior changes.

We've love to get your input on how to move forward. Creating a mechanism to emulate CSS handling might work, but that doesn't feel like the right solution. @WilcoFiers, we'd love to get your thoughts on what might make sense here.

**Comment by straker** [comment]
I'll let @WilcoFiers know you ran into trouble and he'll reach out to you with more directions at a later time (he's currently unavailable)
